logs: add benchmark

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The benchmark reads a JSON log file and measures how long it takes to re-encode
it. The focus is on the encoding of message and values, therefore additional
work (time stamping, caller, writing to file) gets avoided.

The goal is to:
- determine how quickly data can get written in the different formats
- how large the result is

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
logs: add benchmark

The benchmark reads a JSON log file and measures how long it takes to re-encode
it. The focus is on the encoding of message and values, therefore additional
work (time stamping, caller, writing to file) gets avoided.

pohly commented:
Some preliminary results for kubelet on a work during a run with -v5 and 350 tests (from https://testgrid.kubernetes.io/sig-instrumentation-tests#kind-json-logging-master):

```
    benchmark_test.go:91: data/v5/kubelet.log: {totalLines:268566 jsonLines:268551 errorLines:4546 argCounts:map[kref:169306 number:14563 other:19973 string:106904 total:310746]}
BenchmarkLogging/v5/kubelet
BenchmarkLogging/v5/kubelet/stats
BenchmarkLogging/v5/kubelet/stats-36               	1000000000	         0.0000011 ns/op	       0 B/op	       0 allocs/op
BenchmarkLogging/v5/kubelet/printf
BenchmarkLogging/v5/kubelet/printf-36              	       2	 578015074 ns/op	233768564 B/op	 2825670 allocs/op
BenchmarkLogging/v5/kubelet/structured
BenchmarkLogging/v5/kubelet/structured-36          	       1	1212683281 ns/op	600507032 B/op	 4538118 allocs/op
BenchmarkLogging/v5/kubelet/JSON
BenchmarkLogging/v5/kubelet/JSON-36                	       2	 520004898 ns/op	136215908 B/op	 2210431 allocs/op
    benchmark_test.go:144: file sizes: map[JSON:70236539 printf:60504003 structured:61450443]
```

Writing JSON is faster by more than 2x than structured text. It's size is only slightly larger.

I still need to check why there are 4546 lines (`errorLines:4546`) that couldn't be parsed as JSON. There is a non-structured header, but that seems a bit excessive.

I wrote code which detects ObjectRef, which gets used quite a bit (`kref:169306`), more than text strings (`string:106904`). But other structs or slices (`other:19973`) are stored and logged as they come from JSON, which may skew the result a bit because those were handled differently originally in kubelet.

pohly commented:
This works as intended now.

@shivanshu1333 : perhaps you can review it?

pohly committed:
json: more flexible constructor

The encoder configuration can now be chosen by the caller. This will be used by
a benchmark to write messages without caller and time stamp.

While at it, some places where the logger was unnecessarily tested with split
output streams writing into the same actual stream were replaced with writing
as single stream. This is a leftover from a previous incarnation of the split
output stream patch where identical streams were used instead of nil for the
error stream to indicate "single stream".

pohly commented:
@dims: verify-imports fails because k8s.io/component-base/logs now depends on k8s.io/api. That's only for testing. Can we allow that dependency?

pohly commented:
The alternatives are:
- move the new benchmark directory someplace else, for example `k/k/test` (drawback: an entirely new directory that we need to maintain as WG structured logging, test far removed from code)
- add benchmark/go.mod (drawback: `go test ./...` does not include sub-packages)

dims commented:
@pohly please let's not change the order of dependency.

+1 to the first alternative ("move the new benchmark directory someplace else") as the second alternative means one more go.mod file to reconcile which is a pain.

pohly committed:
logs: create separate test/integration directory

The benchmark depends on k8s.io/api (for v1.Container). Such a dependency is
not desirable for k8s.io/component-base/logs, even if it's just for
testing. The solution is to create a separate directory where such a dependency
isn't a problem.

The alternative, a separate package with its own go.mod file under
k8s.io/component-base/logs wouldd have been more complicated to maintain (yet
another go.mod file and different whitelisted dependencies).

pohly commented:
> +1 to the first alternative ("move the new benchmark directory someplace else")

Okay, implemented as `test/integration/logs/benchmark`.

/hold

Let's verify that `pull-kubernetes-verify` covers it.

pohly commented:
/hold cancel

"pull-kubernetes-unit" did not run the benchmark unit tests, but "pull-kubernetes-integration" did. That works for me.

@serathius : agreed? Please review.

pohly commented on test/integration/logs/benchmark/common_test.go:
```diff
@@ -0,0 +1,79 @@
+
+func prints(item logMessage) {
+	if item.isError {
+		klog.ErrorS(item.err, item.msg, item.kvs...)
```
`pull-kubernetes-verify` fails now for the "structured-logging" test: `Additional arguments to ErrorS should always be Key Value pairs. Please check if there is any key or value missing.`

That looks like a bug in the tool. IMHO it should ignore cases that it cannot check statically.

serathius replied:
This was implemented specifically to discourage contributors from providing dynamic keys. I think it should be pretty easy to work around this (create alias for function).

pohly replied:
Is that documented somewhere? Linking to that in the message would have helped. As it stands, that error just looks like a false negative.

I've implemented the workaround.

pohly committed:
logs: work around logcheck

logcheck complains:
Additional arguments to ErrorS should always be Key Value pairs. Please check if there is any key or value missing.

That check is intentional, but not applicable here. The check can be worked
around by calling the functions through variables.

serathius commented on test/integration/logs/benchmark/common_test.go:
```diff
@@ -0,0 +1,79 @@
+var jsonLogger = newJSONLogger(&output)
+
+func newJSONLogger(out io.Writer) logr.Logger {
+	encoderConfig := &zapcore.EncoderConfig{
```
Any reason this config doesn't resemble more the real configuration?

Looks like this skips caller info and uses different encoders than normal JSON formatter. Is this intentional?

serathius replied:
Oh, I didn't review commit by commit so I didn't read whole commit messages.

pohly replied:
To elaborate further, stack unwinding and timestamping adds noise that I didn't want when comparing the encoding aspect. It also makes it hard to verify output.

If we were to enable it, the performance numbers would be questionable because the call site is always the same. I don't know how much that affects performance, but it didn't seem useful.

serathius commented on test/integration/logs/benchmark/common_test.go:
```diff
@@ -0,0 +1,86 @@
+
+func init() {
+	// Cause all klog output to be discarded with minimal overhead.
+	// We don't include time stamps and caller information.
```
Why don't include timestamps and caller information? If we are comparing formats to each other we should make sure to include this overhead as it impacts real world.

pohly commented:
> Oh, I didn't review commit by commit so I didn't read whole commit messages.

My PRs are usually meant to be reviewed commit-by-commit :grin:
- I try to write useful commit messages.
- I try to split up into meaningful smaller chunks where possible.

serathius commented:
Not sure about this goal `The benchmark reads a JSON log file and measures how long it takes to re-encode
it. The focus is on the encoding of message and values, therefore additional
work (time stamping, caller, writing to file) gets avoided.`

Why we are trying to just benchmark encoding and re-encoding. I can understand that encoding caller is a common cost that we cannot do much about so including it makes it harder to compare encoding of logs. However, I would argue that when benchmarking we should include total cost of writing logs. It's true that encoding of always present fields will dominate encoding of edge cases, however measuring real cost will prevent us from making premature optimizations. For example test that only measures encoding of values might show us that klog.KRef is slow, however this doesn't mean we should spend time optimizing this can be 1% of total writing log cost.

If we want to add a benchmark to optimize any encoding we should create a separate benchmark specific to type we are optimizing.

pohly commented:
> Why we are trying to just benchmark encoding and re-encoding. 

Because I needed a way to benchmark the changes in https://github.com/kubernetes/klog/pull/273/commits. Those are all about how the code which does the encoding. Without benchmarks for just those parts I cannot reliably measure whether those changes are an improvement.

> However, I would argue that when benchmarking we should include total cost of writing logs.

Note that these benchmarks print output size. If you want to factor in costs for IO, then looking at the output size already gives a good first estimate - much better than actual measuring it, because that will depend a lot on the hardware and will be noisy.

I bet that if I were to write to disk, I would basically end up measuring disk speed.

> It's true that encoding of always present fields will dominate encoding of edge cases, however measuring real cost will prevent us from making premature optimizations. For example test that only measures encoding of values might show us that klog.KRef is slow, however this doesn't mean we should spend time optimizing this can be 1% of total writing log cost.

For that the benchmark supports encoding of real kubelet and kube-scheduler logs. That gives us a realistic test set where we can see the effect of changes on real-world logs.

> If we want to add a benchmark to optimize any encoding we should create a separate benchmark specific to type we are optimizing.

Do you want different benchmark code or different data files? Those can be added easily.

pohly commented:
How about this:
- we clarify which cases we want to benchmark individually and add the corresponding data files in this PR
- we file enhancement ideas as separate issues and then someone can work on those when time permits

What I don't want is to delay merging this PR because it is useful as-is at least for some purposes.

ehashman commented:
/priority important-longterm
/triage accepted

/assign @serathius

pohly committed:
logs: benchmark write performance

The recent regression https://github.com/kubernetes/kubernetes/issues/107033
shows that we need a way to automatically measure different logging
configurations (structured text, JSON with and without split streams) under
realistic conditions (time stamping, caller identification).

System calls may affect the performance and thus writing into actual files is
useful. A temp dir under /tmp (usually a tmpfs) is used, so the actual IO
bandwidth shouldn't affect the outcome. The "normal" json.Factory code is used
to construct the JSON logger when we have actual files that can be set as
os.Stderr and os.Stdout, thus making this as realistic as possible.

When discarding the output instead of writing it, the focus is more on the rest
of the pipeline and changes there can be investigated more reliably.

The benchmarks automatically gather "log entries per second" and "bytes per
second", which is useful to know when considering requirements like the ones
from https://github.com/kubernetes/kubernetes/issues/107029.

pohly commented:
I added another benchmark that measures writing logs under more realistic scenarios. From the commit message:

> The recent regression https://github.com/kubernetes/kubernetes/issues/107033
> shows that we need a way to automatically measure different logging
> configurations (structured text, JSON with and without split streams) under
> realistic conditions (time stamping, caller identification).
>
> System calls may affect the performance and thus writing into actual files is
> useful. A temp dir under /tmp (usually a tmpfs) is used, so the actual IO
> bandwidth shouldn't affect the outcome. The "normal" json.Factory code is used
> to construct the JSON logger when we have actual files that can be set as
> os.Stderr and os.Stdout, thus making this as realistic as possible.
>
> When discarding the output instead of writing it, the focus is more on the rest
> of the pipeline and changes there can be investigated more reliably.
>
> The benchmarks automatically gather "log entries per second" and "bytes per
> second", which is useful to know when considering requirements like the ones
> from https://github.com/kubernetes/kubernetes/issues/107029.

It's inspired by https://github.com/kubernetes/kubernetes/pull/107032. To fit into the "go test -bench" framework, the test treats writing a certain number of log entries as the stop criteria. `go test` then increases that number automatically to satisfy the intended duration (set with `-benchtime=10s`).

"logs per second" and "bytes per second" are computed automatically and printed as additional information:

```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/logs/benchmark
cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
BenchmarkWriting/discard/structured-36         	      27	 429272951 ns/op
--- BENCH: BenchmarkWriting/discard/structured-36
    benchmark_test.go:266: Wrote 100000 log entries in 427.808492ms -> 233749.5/s
    benchmark_test.go:266: Wrote 2700000 log entries in 11.590542413s -> 232948.5/s
BenchmarkWriting/discard/JSON/single-stream-36 	     214	  55472030 ns/op
--- BENCH: BenchmarkWriting/discard/JSON/single-stream-36
    benchmark_test.go:266: Wrote 100000 log entries in 114.406001ms -> 874080.0/s
    benchmark_test.go:266: Wrote 10000000 log entries in 5.571984675s -> 1794692.7/s
    benchmark_test.go:266: Wrote 21400000 log entries in 11.871087609s -> 1802699.2/s
BenchmarkWriting/discard/JSON/split-stream-36  	     222	  54194252 ns/op
--- BENCH: BenchmarkWriting/discard/JSON/split-stream-36
    benchmark_test.go:266: Wrote 100000 log entries in 77.845286ms -> 1284599.3/s
    benchmark_test.go:266: Wrote 10000000 log entries in 5.389699968s -> 1855390.8/s
    benchmark_test.go:266: Wrote 22200000 log entries in 12.031191837s -> 1845203.7/s
BenchmarkWriting/tmp-files/structured-36       	      12	 969077210 ns/op
--- BENCH: BenchmarkWriting/tmp-files/structured-36
    benchmark_test.go:266: Wrote 100000 log entries in 932.186855ms -> 107274.6/s
    benchmark_test.go:290:       36809900 bytes to file #0 -> 37.7MiB/s (starts with: I1216 15:42:11.267897  197395 benchmark_test.go...)
    benchmark_test.go:266: Wrote 1200000 log entries in 11.629066732s -> 103189.7/s
    benchmark_test.go:290:       441730900 bytes to file #0 -> 36.2MiB/s (starts with: I1216 15:42:12.210798  197395 benchmark_test.go...)
BenchmarkWriting/tmp-files/JSON/single-stream-36         	      15	 740354513 ns/op
--- BENCH: BenchmarkWriting/tmp-files/JSON/single-stream-36
    benchmark_test.go:266: Wrote 100000 log entries in 758.341817ms -> 131866.7/s
    benchmark_test.go:290:       39779465 bytes to file #0 -> 50.0MiB/s (starts with: {"ts":1639665794447.6047,"caller":"benchmark/be...)
    benchmark_test.go:266: Wrote 1500000 log entries in 11.105479068s -> 135068.5/s
    benchmark_test.go:290:       596698790 bytes to file #0 -> 51.2MiB/s (starts with: {"ts":1639665794447.6047,"caller":"benchmark/be...)
BenchmarkWriting/tmp-files/JSON/split-stream-36          	      18	 647593776 ns/op
--- BENCH: BenchmarkWriting/tmp-files/JSON/split-stream-36
    benchmark_test.go:266: Wrote 100000 log entries in 639.584957ms -> 156351.4/s
    benchmark_test.go:290:       716044403 bytes to file #0 -> 1067.7MiB/s (starts with: {"ts":1639665794447.6047,"caller":"benchmark/be...)
    benchmark_test.go:290:       364273 bytes to file #1 -> 0.5MiB/s (starts with: {"ts":1639665856390.5867,"caller":"benchmark/be...)
    benchmark_test.go:266: Wrote 1800000 log entries in 11.656861052s -> 154415.5/s
    benchmark_test.go:290:       716044403 bytes to file #0 -> 58.6MiB/s (starts with: {"ts":1639665794447.6047,"caller":"benchmark/be...)
    benchmark_test.go:290:       7244711 bytes to file #1 -> 0.6MiB/s (starts with: {"ts":1639665856390.5867,"caller":"benchmark/be...)
```

There's just one problem with it... I can't reproduce the performance regression from https://github.com/kubernetes/kubernetes/issues/107033 on my machine. Applying this patch on top of this PR should trigger the regression (error and info stream synced again, non-split-stream accidentally uses split-stream):
```patch
diff --git a/staging/src/k8s.io/component-base/logs/json/json.go b/staging/src/k8s.io/component-base/logs/json/json.go
index 1adada10cd4..254fd9ab92f 100644
--- a/staging/src/k8s.io/component-base/logs/json/json.go
+++ b/staging/src/k8s.io/component-base/logs/json/json.go
@@ -54,6 +54,18 @@ func NewJSONLogger(infoStream, errorStream zapcore.WriteSyncer, encoderConfig *z
 	if errorStream == nil {
 		core = zapcore.NewCore(encoder, infoStream, zapcore.Level(-127))
 	} else {
+		// Set up writing of error messages to stderr and info messages
+		// to stdout. Info messages get optionally buffered and flushed
+		// - through klog.FlushLogs -> zapr Flush -> zap Sync
+		// - when an error gets logged
+		//
+		// The later is important when both streams get merged into a single
+		// stream by the consumer (same console for a command line tool, pod
+		// log for a container) because without it, messages get reordered.
+		flushError := writeWithFlushing{
+			WriteSyncer: errorStream,
+			other:       infoStream,
+		}
 		highPriority := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
 			return lvl >= zapcore.ErrorLevel
 		})
@@ -61,7 +73,7 @@ func NewJSONLogger(infoStream, errorStream zapcore.WriteSyncer, encoderConfig *z
 			return lvl < zapcore.ErrorLevel
 		})
 		core = zapcore.NewTee(
-			zapcore.NewCore(encoder, errorStream, highPriority),
+			zapcore.NewCore(encoder, flushError, highPriority),
 			zapcore.NewCore(encoder, infoStream, lowPriority),
 		)
 	}
@@ -101,5 +113,17 @@ func (f Factory) Create(options config.FormatOptions) (logr.Logger, func()) {
 		return NewJSONLogger(stdout, stderr, nil)
 	}
 	// Write info messages and errors to stderr to prevent mixing with normal program output.
-	return NewJSONLogger(stderr, nil, nil)
+	return NewJSONLogger(stderr, stderr, nil)
+}
+
+// writeWithFlushing is a wrapper around an output stream which flushes another
+// output stream before each write.
+type writeWithFlushing struct {
+	zapcore.WriteSyncer
+	other zapcore.WriteSyncer
+}
+
+func (f writeWithFlushing) Write(bs []byte) (int, error) {
+	f.other.Sync()
+	return f.WriteSyncer.Write(bs)
 }
```

But the results for `go test -timeout=0 -bench=BenchmarkWriting -benchtime=10s -count=5 ./test/integration/logs/benchmark` aren't very different:
```
$ benchstat /tmp/master /tmp/regression 
name                                     old time/op  new time/op  delta
Writing/discard/structured-36             405ms ± 6%   362ms ±29%    ~     (p=0.056 n=5+5)
Writing/discard/JSON/single-stream-36    54.6ms ± 3%  54.7ms ± 3%    ~     (p=1.000 n=5+5)
Writing/discard/JSON/split-stream-36     54.5ms ± 2%  54.2ms ± 3%    ~     (p=0.548 n=5+5)
Writing/tmp-files/structured-36           969ms ± 0%   978ms ± 1%  +0.87%  (p=0.016 n=4+5)
Writing/tmp-files/JSON/single-stream-36   675ms ±10%   692ms ±10%    ~     (p=0.310 n=5+5)
Writing/tmp-files/JSON/split-stream-36    645ms ± 0%   659ms ± 1%  +2.23%  (p=0.008 n=5+5)
```

I'll try with the benchmark from https://github.com/kubernetes/kubernetes/pull/107032. If that also shows no difference for me, then my attempt to re-introduce the flaw is flawed or my machine isn't affected.

serathius commented:
Great work, in my case as /tmp is not tmpfs I see the regression. We could switch it to real disk, however effects could greatly depend on hardware, maybe instead we could introduce a throughput limited sink that simulates characteristics of disks we expect users to use like in https://github.com/kubernetes/kubernetes/issues/107029#issuecomment-995699791

pohly commented:
When I run with `TMPDIR=/backup` where `/backup` are real HDs, I see the regression also with this benchmark:
```
$ $GOPATH/bin/benchstat /tmp/master /tmp/regression 
name                                     old time/op  new time/op   delta
Writing/discard/structured-36             413ms ± 4%    418ms ± 4%   ~     (p=0.700 n=3+3)
Writing/discard/JSON/single-stream-36    59.0ms ± 4%   57.2ms ± 4%   ~     (p=0.400 n=3+3)
Writing/discard/JSON/split-stream-36     56.8ms ± 2%   55.6ms ± 1%   ~     (p=0.200 n=3+3)
Writing/tmp-files/structured-36           1.39s ± 0%    1.43s ± 5%   ~     (p=0.200 n=3+3)
Writing/tmp-files/JSON/single-stream-36   931ms ±17%  52295ms ±12%   ~     (p=0.100 n=3+3)
Writing/tmp-files/JSON/split-stream-36    816ms ± 1%  48589ms ± 0%   ~     (p=0.100 n=3+3)
```

The variance is very high, though, therefore `benchstat` didn't report a delta.

pohly commented:
> maybe instead we could introduce a throughput limited sink that simulates characteristics of disks we expect users to use

That should work. We would have to implement the `Sync` method for a fake in-memory `SyncWriter`. But what exactly are those characteristics?

Should `Sync` always take a certain fixed amount of time, or should it depend on the amount of data since the last sync? Perhaps the answer is very simple: our logging stack should never call it in the first place! Syncing to disk is only useful if the node is about to crash. I think we don't need to worry about that. If true, then our "fake Sync" can take an arbitrary, very long time or even crash the test when called.

More important is the average write rate. But that then begs the question whether short bursts are allowed.

dims commented:
Do we want to land this in 1.24 cycle @pohly ?

pohly commented:
Yes, let's target 1.24.

serathius commented:
> When I run with `TMPDIR=/backup` where `/backup` are real HDs, I see the regression also with this benchmark:
> 
> ```
> $ $GOPATH/bin/benchstat /tmp/master /tmp/regression 
> name                                     old time/op  new time/op   delta
> Writing/discard/structured-36             413ms ± 4%    418ms ± 4%   ~     (p=0.700 n=3+3)
> Writing/discard/JSON/single-stream-36    59.0ms ± 4%   57.2ms ± 4%   ~     (p=0.400 n=3+3)
> Writing/discard/JSON/split-stream-36     56.8ms ± 2%   55.6ms ± 1%   ~     (p=0.200 n=3+3)
> Writing/tmp-files/structured-36           1.39s ± 0%    1.43s ± 5%   ~     (p=0.200 n=3+3)
> Writing/tmp-files/JSON/single-stream-36   931ms ±17%  52295ms ±12%   ~     (p=0.100 n=3+3)
> Writing/tmp-files/JSON/split-stream-36    816ms ± 1%  48589ms ± 0%   ~     (p=0.100 n=3+3)
> ```
> 
> The variance is very high, though, therefore `benchstat` didn't report a delta.

You need to increase test count, `3` is pretty low. I usually run `5` or `10` for any benchmark. 

> > maybe instead we could introduce a throughput limited sink that simulates characteristics of disks we expect users to use
> 
> That should work. We would have to implement the `Sync` method for a fake in-memory `SyncWriter`. But what exactly are those characteristics?
> 
> Should `Sync` always take a certain fixed amount of time, or should it depend on the amount of data since the last sync? Perhaps the answer is very simple: our logging stack should never call it in the first place! Syncing to disk is only useful if the node is about to crash. I think we don't need to worry about that. If true, then our "fake Sync" can take an arbitrary, very long time or even crash the test when called.
> 
> More important is the average write rate. But that then begs the question whether short bursts are allowed.

At the moment I don't have good intuition for this. Main goal of simulating sinks like disk would be to make results easier to reproduce, by eliminating hardware dependence.  I think we should try to have some simple model based on real measurements like in https://github.com/kubernetes/kubernetes/issues/107029. With measurements done on different cloud providers we could create a list of hardcoded sinks. 

Results presented by @pohly show that we can reproduce the regression allowing us to prevent such cases in the future. I think we can move forward and merge this. Further improvements should be done in separate PR.

serathius commented:
/lgtm

serathius commented:
/assign @lavalamp

pohly commented:
/hold

The new test/integration/logs directory should have an OWNERS file which lists wg-structured logging.

serathius commented:
FYI working groups should not own code by itself. We could list contributors and SIG Instrumentation as OWNERS.

pohly committed:
json: never call fsync for stdout or stderr

We don't need to worry about data loss once the data has been written to an
output stream. Calling fsync unnecessarily has been the reason for performance
issues in the past.

pohly commented:
/hold cancel

I've added the same OWNERS file as under component-base/logs, i.e. SIG-Instrumentation owns the code, with @serathius and myself as approvers because we are the WG leads.

I also added one commit which suppresses fsync, as discussed on Slack.

@serathius: can you take another look and re-add the LGTM?

pohly committed:
logs: update expected output for multi-line test case

The multiline support has been merged while this benchmark was written. We now
get the output that we want, with line breaks.

serathius commented:
Question about disabling sync. I have looked into previous discussion about this in zap project. https://github.com/uber-go/zap/issues/370. Looks like both OSX and Linux doesn't support fsync on stdout&stderr. 

In that case is there any need to disable sync if we are directly operating on stdout and stderr?

pohly commented:
> Question about disabling sync. I have looked into previous discussion about this in zap project. [uber-go/zap#370](https://github.com/uber-go/zap/issues/370). Looks like both OSX and Linux doesn't support fsync on stdout&stderr.

That issue probably refers to the case where stdout is /dev/pts or a pipe. Then fsync returns an error or at least has no effect.

> In that case is there any need to disable sync if we are directly operating on stdout and stderr?

stdout and stderr might also be a real file (`kubelet 2>/var/log/kubelet.log >&2`). In that case fsync works.

pohly commented:
/assign @lavalamp 

For approval. We need a new directory to host the testing code from SIG-Instrumentation/WG structured logging.

pohly commented:
/assign @thockin 

For approval.

Looks like @lavalamp is not around at the moment. See comment above for the explanation why this needs approval from outside the current SIG-Instrumentation team.

thockin reviewed (commented):
Thanks!

/lgtm
/approve

pohly commented:
/retest
